### PR TITLE
[WIP] Add encapsulated `GroupBy` and basic aggregate functions.

### LINKED ISSUE #181: `groupby` method problem
Hi, I'm very appreciate of your nice work !

Here I found a problem of `groupby` function between `pandas.DataFrame` and `koalas.DataFrame`.

```python
import databricks.koalas as ks
import pandas as pd

# Create a pandas DataFrame
pdf = pd.DataFrame({'x':range(3), 'y':['a','b','b'], 'z':['a','b','b']})
# Create a Koalas DataFrame from pandas DataFrame
kdf = ks.from_pandas(pdf)

# pandas groupby, works well
print(pdf.groupby('y')['x'].mean())
## y
## a    0.0
## b    1.5
## Name: x, dtype: float64
```
However, when using `koalas`,
```python
# koalas groupby, throw error
print(kdf.groupby('y')['x'].mean())
```
```shell
ValueError                                Traceback (most recent call last)
<ipython-input-71-1d6e96373165> in <module>()
      3
      4 # koalas groupby, throw error
----> 5 print(kdf.groupby('y')['x'].mean())

/usr/install/anaconda3/lib/python3.6/site-packages/databricks/koalas/groups.py in __getitem__(self, key)
     49             if k not in fnames:
     50                 raise ValueError("{} not a field. Possible values are {}".format(k, fnames))
---> 51         return PandasLikeGroupBy(self._df, self._groupdata, key)
     52
     53     def __getattr__(self, key):

/usr/install/anaconda3/lib/python3.6/site-packages/databricks/koalas/groups.py in __getattr__(self, key)
     53     def __getattr__(self, key):
     54         try:
---> 55             return self[key]
     56         except KeyError as e:
     57             raise AttributeError(e)

/usr/install/anaconda3/lib/python3.6/site-packages/databricks/koalas/groups.py in __getitem__(self, key)
     48         for k in l:
     49             if k not in fnames:
---> 50                 raise ValueError("{} not a field. Possible values are {}".format(k, fnames))
     51         return PandasLikeGroupBy(self._df, self._groupdata, key)
     52

ValueError: _df not a field. Possible values are {'y', '__index_level_0__', 'x', 'z'}
```
<br/>

My environment:
- python 3.6.5
- spark 2.4.0

### PR DESCRIPTION
**Pull request by ueshin**
Switch over to the encapsulation approach for `groupby`, and add basic aggregate functions.
In this PR, only `by` argument will be supported.

The aggregate functions which will be addressed in this PR are:

- `agg`/`aggregate`
- `count`
- `first`
- `last`
- `max`
- `mean`
- `min`
- `std`
- `sum`
- `var`

TODOs:
- [ ] add `agg`/`aggregate` function
- [ ] docs

Resolves #181, #185.

### CODE REVIEW

File: databricks/koalas/generic.py [hunk 1]
```diff
@@ -84,6 +84,19 @@ def abs(self):
         """
         return _spark_col_apply(self, F.abs)
 
+    def groupby(self, by):
```
Comment (rxin) [inline databricks/koalas/generic.py]: for later: add docstring (can just copy paste Pandas')

File: databricks/koalas/groupby.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+class GroupBy(object):
+
+    def __new__(cls, obj: Union[DataFrame, Series], *args, **kwargs):
+        if isinstance(obj, DataFrame):
+            return super(GroupBy, cls).__new__(DataFrameGroupBy)
+        elif isinstance(obj, Series):
+            return super(GroupBy, cls).__new__(SeriesGroupBy)
+        else:
+            raise TypeError('invalid type: {}'.format(obj))
```
Comment (rxin) [inline databricks/koalas/groupby.py]: in addition to pointing out it is invalid, say we only allow DataFrame or Series.

File: databricks/koalas/testing/utils.py [hunk 3]
```diff
@@ -194,30 +194,44 @@ def assertPandasAlmostEqual(self, left, right):
...
+            self.assertEqual(len(left), len(right), msg=msg)
+            for lnull, rnull in zip(left.isnull(), right.isnull()):
+                self.assertEqual(lnull, rnull, msg=msg)
+            for lval, rval in zip(left.dropna(), right.dropna()):
                 self.assertAlmostEqual(lval, rval, msg=msg)
         else:
             raise ValueError("Unexpected values: (%s, %s)" % (left, right))
 
-    def assert_eq(self, left, right):
+    def assert_eq(self, left, right, almost=False):
```
Comment (rxin) [inline databricks/koalas/testing/utils.py]: while you at it, document what the almost argument.

File: databricks/koalas/generic.py [hunk 4]
```diff
@@ -84,6 +84,19 @@ def abs(self):
...
+        from databricks.koalas.groupby import GroupBy
+        from databricks.koalas.series import Series
+        if isinstance(by, str):
+            by = [by]
+        elif isinstance(by, Series):
```
Comment (thunterdb) [inline databricks/koalas/generic.py]: check the anchor of the series is the same too.

File: databricks/koalas/groupby.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    _MissingPandasLikeSeriesGroupBy
+from databricks.koalas.series import Series, _col
+
+
+ColumnLike = Union[str, Series]
+
+
+class GroupBy(object):
+
+    def __new__(cls, obj: Union[DataFrame, Series], *args, **kwargs):
```
Comment (thunterdb) [inline databricks/koalas/groupby.py]: do you really have to make it the union of either? You should pick he most appropriate

File: databricks/koalas/generic.py (hunk 4 above)
Comment (thunterdb) [reply databricks/koalas/generic.py]: also, you should split that ino the SeriesGroupBy and the DataFrameGroupBy into `frames.py` and `series.py`?

File: databricks/koalas/missing/groupby.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+#
+
+from databricks.koalas.missing import _unsupported_function
+
+
+def unsupported_function(method_name):
+    return _unsupported_function(class_name='pd.groupby.GroupBy', method_name=method_name)
+
+
+class _MissingPandasLikeDataFrameGroupBy(object):
```
Comment (thunterdb) [inline databricks/koalas/missing/groupby.py]: There is all that in Pandas groupby? I never realized

File: databricks/koalas/generic.py (hunk 4 above)
Comment (HyukjinKwon) [reply databricks/koalas/generic.py]: I'll split them when we actually see considerable divergence between them.

**Pull request closed without merging**